### PR TITLE
move api/v2/version endpoint to OPTIONS /api

### PR DESCRIPTION
**Pull request by bartcharbon**
#### Checklist
- [x] Functionality works & meets specifications
- [x] Code reviewed
- [x] Code unit/integration/system tested
- [n.a.] Migration step added in case of breaking change
- [n.a.] User documentation updated
- [x] (If you have changed REST API interface) view-swagger.ftl updated
- [n.a.] Test plan template updated
- [x] Clean commits
- [x] Added Feature/Fix to release notes

### CODE REVIEW

File: molgenis-api-data/src/main/java/org/molgenis/api/data/v2/RestControllerV2.java [hunk 1]
```diff
@@ -142,29 +145,27 @@ public RestControllerV2(
...
     this.permissionService = requireNonNull(permissionService);
     this.restService = requireNonNull(restService);
     this.localizationService = requireNonNull(localizationService);
     this.permissionSystemService = requireNonNull(permissionSystemService);
     this.repoCopier = requireNonNull(repoCopier);
+    this.versionController = requireNonNull(versionController);
   }
 
   @Autowired
+  @Deprecated
```
Comment (dennishendriksen) [inline molgenis-api-data/src/main/java/org/molgenis/api/data/v2/RestControllerV2.java]: I suggest that the only change to this controller is adding the @Deprecated annotation as well as a @deprecated javadoc annotation that refers to the preferred method.

File: molgenis-api-data/src/test/java/org/molgenis/api/data/v2/RestControllerV2Test.java [hunk 2]
```diff
@@ -55,6 +55,7 @@
 import static org.testng.reporters.Files.readFile;
 
 import com.google.common.collect.Sets;
+import java.text.ParseException;
```
Comment (dennishendriksen) [inline molgenis-api-data/src/test/java/org/molgenis/api/data/v2/RestControllerV2Test.java]: Is suggest not to change any tests (see comment in RestControllerV2).

File: molgenis-api/src/main/java/org/molgenis/api/ApiController.java [hunk 3]
```diff
@@ -17,7 +17,7 @@
    */
   public ApiController(String apiId, Integer apiVersion) {
     requireNonNull(apiId);
-    if (!API_ID_PATTERN.matcher(apiId).matches()) {
+    if (!apiId.isEmpty() && !API_ID_PATTERN.matcher(apiId).matches()) {
```
Comment (dennishendriksen) [inline molgenis-api/src/main/java/org/molgenis/api/ApiController.java]: Add a unit test for empty apiId

File: molgenis-api/src/main/java/org/molgenis/api/VersionApiNamespace.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package org.molgenis.api;
+
+public class VersionApiNamespace {
```
Comment (dennishendriksen) [inline molgenis-api/src/main/java/org/molgenis/api/VersionApiNamespace.java]: We agreed on ```OPTIONS /api``` to expose the application version, right? That would make this class unnecessary.

File: molgenis-api/src/main/java/org/molgenis/api/VersionController.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import java.text.SimpleDateFormat;
+import java.time.Instant;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping(API_PATH)
+public class VersionController extends ApiController {
```
Comment (dennishendriksen) [inline molgenis-api/src/main/java/org/molgenis/api/VersionController.java]: How about renaming this class to RootApiController so we can extend it with other planned endpoints such as listing all apis with their versions?

File: molgenis-api/src/main/java/org/molgenis/api/VersionController.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+      @Value("${molgenis.version:@null}") String molgenisVersion,
+      @Value("${molgenis.build.date:@null}") String molgenisBuildDate) {
+    super("", 1);
+    this.molgenisVersion = requireNonNull(molgenisVersion);
+    this.molgenisBuildDate = requireNonNull(molgenisBuildDate);
+  }
+
+  @Autowired
+  @RequestMapping(method = OPTIONS)
+  public VersionResponse getVersion() throws ParseException {
```
Comment (dennishendriksen) [inline molgenis-api/src/main/java/org/molgenis/api/VersionController.java]: I suggest storing the version information under the app key.
```json
{
  "app": {
    "version": "...",
    "buildDate": "..."
  },
  "apis": []
}
```
The apis key could be added in the future.

File: molgenis-api/src/test/java/org/molgenis/api/VersionControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  @Test
+  public void testGetVersion() throws Exception {
+    MvcResult result =
+        mockMvc
+            .perform(options("/api").contentType(APPLICATION_JSON_UTF8))
+            .andExpect(status().isOk())
+            .andReturn();
+    String actual = result.getResponse().getContentAsString();
+    String expected =
+        "{\"molgenisVersion\":\"versionString\",\"buildDate\":\"2019-05-21T10:32:00Z\"}";
```
Comment (dennishendriksen) [inline molgenis-api/src/test/java/org/molgenis/api/VersionControllerTest.java]: I think that it is a good idea to use one format for date-times in JSON requests/responses everywhere. Did you look into adding milliseconds? It would make sense considering the places where we use datetimes.

File: molgenis-api/src/main/java/org/molgenis/api/VersionResponse.java [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+import com.google.auto.value.AutoValue;
+import java.time.Instant;
+import org.molgenis.util.AutoGson;
+
+@AutoValue
+@AutoGson(autoValueClass = AutoValue_VersionResponse.class)
+@SuppressWarnings("squid:S1610")
+public abstract class VersionResponse {
+  public abstract String getMolgenisVersion();
```
Comment (dennishendriksen) [inline molgenis-api/src/main/java/org/molgenis/api/VersionResponse.java]: ```getVersion``` instead of ```getMolgenisVersion```

File: molgenis-api/src/test/java/org/molgenis/api/VersionControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+  private MockMvc mockMvc;
+
+  @BeforeMethod
+  public void beforeMethod() {
+    VersionController versionController =
+        new VersionController("versionString", "2019-05-21 10:32 UTC");
+    mockMvc =
+        MockMvcBuilders.standaloneSetup(versionController)
+            .setMessageConverters(new FormHttpMessageConverter(), gsonHttpMessageConverter)
```
Comment (dennishendriksen) [inline molgenis-api/src/test/java/org/molgenis/api/VersionControllerTest.java]: I suspect you don't need a FormHttpMessageConverter here.

File: molgenis-api/src/test/java/org/molgenis/api/VersionControllerTest.java (hunk 7 above)
Comment (bartcharbon) [reply molgenis-api/src/test/java/org/molgenis/api/VersionControllerTest.java]: Response object contains an Instant, the conversion to json is than done centrally by the application right?

File: molgenis-api/src/test/java/org/molgenis/api/VersionControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+@TestExecutionListeners(listeners = WithSecurityContextTestExecutionListener.class)
+@TestPropertySource(properties = {"molgenis.version = 10.3.8"})
+@ContextConfiguration(classes = {GsonConfig.class})
+public class VersionControllerTest extends AbstractTestNGSpringContextTests {
```
Comment (dennishendriksen) [inline molgenis-api/src/test/java/org/molgenis/api/VersionControllerTest.java]: Rename to RootApiControllerTest

File: molgenis-api/src/main/java/org/molgenis/api/RootApiController.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+import java.text.SimpleDateFormat;
+import java.time.Instant;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping(API_PATH)
+public class RootApiController extends ApiController {
```
Comment (dennishendriksen) [inline molgenis-api/src/main/java/org/molgenis/api/RootApiController.java]: Add swagger annotations

File: molgenis-api/src/main/java/org/molgenis/api/VersionResponse.java [hunk 12]
```diff
@@ -0,0 +1,15 @@
+package org.molgenis.api;
+
+import com.google.auto.value.AutoValue;
+import org.molgenis.util.AutoGson;
+
+@AutoValue
+@AutoGson(autoValueClass = AutoValue_VersionResponse.class)
+@SuppressWarnings("squid:S1610")
+public abstract class VersionResponse {
```
Comment (dennishendriksen) [inline molgenis-api/src/main/java/org/molgenis/api/VersionResponse.java]: Rename to something like OptionsResponse/ApiOptionsResponse

File: molgenis-api/src/main/java/org/molgenis/api/RootApiController.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+      @Value("${molgenis.version:@null}") String molgenisVersion,
+      @Value("${molgenis.build.date:@null}") String molgenisBuildDate) {
+    super("", 1);
+    this.molgenisVersion = requireNonNull(molgenisVersion);
+    this.molgenisBuildDate = requireNonNull(molgenisBuildDate);
+  }
+
+  @Autowired
+  @RequestMapping(method = OPTIONS)
+  public VersionResponse getVersion() throws ParseException {
```
Comment (dennishendriksen) [inline molgenis-api/src/main/java/org/molgenis/api/RootApiController.java]: Deprecate /api/v2 version endpoint and refer to this endpoint

File: molgenis-api/src/main/java/org/molgenis/api/RootApiController.java (hunk 11 above)
Comment (dennishendriksen) [reply molgenis-api/src/main/java/org/molgenis/api/RootApiController.java]: swagger not working yet
